tools/CMakeLists.txt

```
@@ -82,7 +82,7 @@ endif(Python_EXECUTABLE AND NOT PYTHON_INSTDIR)
 # Checks what linear algebra packages are installed                    #
 ########################################################################
 
-if(CMAKE_CXX_COMPILER_ID STREQUAL "Intel")
+if(CMAKE_CXX_COMPILER_ID STREQUAL "Intel" OR CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
```

```suggestion
if(CMAKE_CXX_COMPILER_ID MATCHES "Intel" OR CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
```
@baumeier does that help?

---

I believe `dpcpp` identities itself as `IntelLLVM`. Can you confirm by printing `CMAKE_CXX_COMPILER_ID` on your system?

---

We may need to think about `dpcpp` separately. Depending on CMake version, it is either detected as `Clang` (without Intel) and in newer versions as `IntelLLVM`...

---

CMake 3.18.4
` The CXX compiler identification is Clang 13.0.0`

CMake 3.18.4
`-- The CXX compiler identification is IntelLLVM 2021.4.0`

---

Wouldn't the above then work for either case?

---

OK, here is the overview of all the `CMake` outputs (selected):

1. Current Master
With `CMake 3.18.4`

```
-- The CXX compiler identification is Clang 13.0.0
[...]
CMake Error at /cm/shared/apps/cmake/3.18.4/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:165 (message):
  Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES)
Call Stack (most recent call first):
  /cm/shared/apps/cmake/3.18.4/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:458 (_FPHSA_FAILURE_MESSAGE)
  /cm/shared/apps/cmake/3.18.4/share/cmake-3.18/Modules/FindOpenMP.cmake:529 (find_package_handle_standard_args)
  CMakeLists.txt:114 (_find_package)
  tools/CMakeModules/FindMKL.cmake:324 (find_package)
  CMakeLists.txt:114 (_find_package)
  tools/CMakeLists.txt:90 (find_package)
```

With `CMake 3.21.4`

```
-- The CXX compiler identification is IntelLLVM 2021.4.0
[...]
CMake Error at /home/mcs001/bbaumeie/ownlibs/cmake-3.21.4-linux-x86_64/share/cmake-3.21/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES)
Call Stack (most recent call first):
  /home/mcs001/bbaumeie/ownlibs/cmake-3.21.4-linux-x86_64/share/cmake-3.21/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  /home/mcs001/bbaumeie/ownlibs/cmake-3.21.4-linux-x86_64/share/cmake-3.21/Modules/FindOpenMP.cmake:544 (find_package_handle_standard_args)
  CMakeLists.txt:114 (_find_package)
  tools/CMakeModules/FindMKL.cmake:324 (find_package)
```

2. This branch with `if(CMAKE_CXX_COMPILER_ID STREQUAL "Intel" OR CMAKE_CXX_COMPILER_ID STREQUAL "Clang")`

With `CMake 3.18.4`

```
-- The CXX compiler identification is Clang 13.0.0
[...]
-- MKL: Thread Layer(Intel OpenMP) Interface(4-byte Integer)
-- Found MKL: /home/mcs001/bbaumeie/intel/oneapi/mkl/2021.4.0/include (found version "2021.0.4")
-- Found MKL_Static: /home/mcs001/bbaumeie/intel/oneapi/mkl/2021.4.0/include (found version "2021.0.4")
-- Found Eigen3: /home/mcs001/bbaumeie/ownlibs/eigen/share/eigen3/cmake
[...]
-- The following OPTIONAL packages have not been found:
 * OpenMP
```

With `CMake 3.21.4`

```
-- The CXX compiler identification is IntelLLVM 2021.4.0
[...]
CMake Error at /home/mcs001/bbaumeie/ownlibs/cmake-3.21.4-linux-x86_64/share/cmake-3.21/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES)
Call Stack (most recent call first):
  /home/mcs001/bbaumeie/ownlibs/cmake-3.21.4-linux-x86_64/share/cmake-3.21/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  /home/mcs001/bbaumeie/ownlibs/cmake-3.21.4-linux-x86_64/share/cmake-3.21/Modules/FindOpenMP.cmake:544 (find_package_handle_standard_args)
  CMakeLists.txt:114 (_find_package)
  tools/CMakeModules/FindMKL.cmake:328 (find_package)
  CMakeLists.txt:114 (_find_package)
  tools/CMakeLists.txt:90 (find_package)
```
 
3. This branch with `if(CMAKE_CXX_COMPILER_ID MATCHES "Intel" OR CMAKE_CXX_COMPILER_ID STREQUAL "Clang")`

With `CMake 3.18.4`

Same as 2.

With `CMake 3.21.4`

```
-- The CXX compiler identification is IntelLLVM 2021.4.0
[...]
-- MKL: Thread Layer(Intel OpenMP) Interface(4-byte Integer)
-- Found MKL: /home/mcs001/bbaumeie/intel/oneapi/mkl/2021.4.0/include (found version "2021.0.4")
-- Found MKL_Static: /home/mcs001/bbaumeie/intel/oneapi/mkl/2021.4.0/include (found version "2021.0.4")
-- Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES) (Required is at least version "4.5")
-- Could NOT find OpenMP (missing: OpenMP_CXX_FOUND) (Required is at least version "4.5") 
[...]
-- The following RECOMMENDED packages have not been found:
 * OpenMP (required version >= 4.5)
```

---

So, indeed `MATCHES` may be a good idea but I am confused about this `OpenMP` finding issue.

---

For the `OneAPI` version of `icc` everything is found as in `CMakeCache.txt`

```
//The thread layer to choose for MKL
MKL_THREAD_LAYER:STRING=Intel OpenMP

//Path to a library.
MKL_ThreadLayer_LINK_LIBRARY:FILEPATH=/opt/intel/mkl/lib/intel64/libmkl_intel_thread.so

//Path to a library.
MKL_ThreadLayer_STATIC_LINK_LIBRARY:FILEPATH=/opt/intel/mkl/lib/intel64/libmkl_intel_thread.a

//Path to a library.
MKL_ThreadingLibrary_LINK_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libiomp5.so

[...]

//CXX compiler flags for OpenMP parallelization
OpenMP_CXX_FLAGS:STRING=-qopenmp
//CXX compiler libraries for OpenMP parallelization
OpenMP_CXX_LIB_NAMES:STRING=iomp5;pthread
//Path to the iomp5 library for OpenMP
OpenMP_iomp5_LIBRARY:FILEPATH=/opt/intel/oneapi/compiler/2021.4.0/linux/compiler/lib/intel64_lin/libiomp5.so
//Path to the pthread library for OpenMP
OpenMP_pthread_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libpthread.so
```

Should we be concerned that there are two different `libiomp5.so` referenced here?

With `dpcpp` on the other hand:

```
//The thread layer to choose for MKL
MKL_THREAD_LAYER:STRING=Intel OpenMP

//Path to a library.
MKL_ThreadLayer_LINK_LIBRARY:FILEPATH=/opt/intel/mkl/lib/intel64/libmkl_intel_thread.so

//Path to a library.
MKL_ThreadLayer_STATIC_LINK_LIBRARY:FILEPATH=/opt/intel/mkl/lib/intel64/libmkl_intel_thread.a

//Path to a library.
MKL_ThreadingLibrary_LINK_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libiomp5.so

[...]

//CXX compiler flags for OpenMP parallelization
OpenMP_CXX_FLAGS:STRING=NOTFOUND

//CXX compiler libraries for OpenMP parallelization
OpenMP_CXX_LIB_NAMES:STRING=NOTFOUND


```

---

It gets even better, now that I realized that not only is there the old `icc` but also the new `icpx` and this data-optimized `dpcpp`...

I guess we should focus on `icpx`. The annoying thing is that it is identified as the same compiler as `dpcpp` running `cmake` (`clang 13.0.0` with `3.18.4` and `IntelLLVM` in 3.21.4). However, it is only built in serial with `cmake 3.18.4` while in parallel only with `3.21.4`.

I have no clue what we should do here, require newer `cmake` for `icpx`?

---

Or maybe we should leave this out completely, as

> In short, if you are looking for performance on CPU targets, ICC/ICPC compiler would be an ideal choice. For OpenMP GPU offload support, you may rely on ICX/ICPX compiler. For single-source heterogenous cross-compilation (that can target CPUs, GPUs, FPGAs, etc), DPC++ is the best choice.

---

I think we should just enforce a newer CMake version, something like:
https://github.com/votca/votca/blob/df08234f9f738e0b546b3d016b0d68b077e885f0/CMakeLists.txt#L32-L34

---

Not sure how this would work, if the Compiler is not ID'd as `Intel` in the old `cmake` version in the first place?

---

Maybe via checking the `CMAKE_CXX_COMPILER` executable explicitly? Then we could also throw an error when someone uses `dpcpp`

---

Yeah, that sounds good, something like:
```cmake
if (CMAKE_CXX_COMPILER MATCHES "dpcpp" AND CMAKE_VERSION VERSION_LESS "3.21") 
   message(FATAL_ERROR "Using the intel dpcpp compiler requires cmake 3.21 or higher") 
 endif()
```

---

And then, I still don't think that with `3.18` at least, we manage to build with OpenMP using `icpx` due to the `clang` ID.